Fix for Issue 4329: Updated dropzone provider to check hovered zones on dragover.

chriskmnds opened the pull request:
## Description
<!-- Please describe your changes -->
This is a fix for issue [4329](https://github.com/WordPress/gutenberg/issues/4329).

To my understanding, there is an issue with the top/bottom dropzones. Initially the dropzones were checked and activated on the dragenter and dragleave events - which fire once each, when we enter an area while dragging and when we leave that area. So when we are already dragging somewhere in the bottom part of the top dropzone and drag the item towards the top of it - the positioning is never updated (since nothing keeps track of the cursor position), and once we drag passed the dropzone then we are not within the zone bounds for the update logic to fire - hence we need to come back into the dropzone area to re-activate the top/first dropzone. 

With `dragover` (the fix), we keep track of the cursor position and do our checks constantly + as this was written in a way that only the dropzones that need to be updated will change their state, then the overhead is not severe. This way we have the update logic fire when we drag from the bottom half to the top half of the first dropzone (and vice versa for the bottom one), hence the activation of the topmost and bottommost goes through as we are on our way to leave the dropzone area.

### Another Approach

**Another possibility** to investigate for a different solution may be to pad the top part of the blocks list with an additional dropzone in order to have the `dragenter` handler fire when leaving the topmost dropzone (so we never really leave the dropzone area, even though we won't be hovering over one of the visible blocks). This may not require changes to the provider.

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Manually. Dragging an image from the desktop into a block and image gallery dropzone - confirming insert.

## Screenshots (jpeg or gifs if applicable):

A gif has been added to the issue referenced above.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Updated handler to check on dragover instead of once on dragenter/leave. Reasoning: we need to tell when the cursor goes from the bottom to the top of the same block, otherwise we can't activate the top/bottom dropzones in the block list unless we drag something out and into the top/bottom dropzones respectively.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

chriskmnds committed:
Fix for Issue 4329: Updated dropzone provider to check hovered zone
boundaries on dragover instead of dragenter/leave.

chriskmnds committed:
Fix for Issue 4329: Linting errors.

chriskmnds committed:
Fix for Issue 4329: Cleanup.

nb commented on components/drop-zone/provider.js:
```diff
@@ -92,21 +83,29 @@ class DropZoneProvider extends Component {
 		const { onVerifyValidTransfer } = this.props;
 		const isValidDrag = ! onVerifyValidTransfer || onVerifyValidTransfer( detail.dataTransfer );
+
+		// Always true (onVerifyValidTransfer is not defined).
```
Can we remove it if it's always `true`?

chriskmnds replied:
We can, although I can't tell what the original intention was to have this property in the first place. For the moment the property is not defined. Maybe @youknowriad can give us a bit more insight.

youknowriad replied:
I'm fine removing this, this is a leftover since this component has been adapted from Calypso

nb commented on components/drop-zone/provider.js:
```diff
@@ -76,10 +71,6 @@ class DropZoneProvider extends Component {
 	}
 
-	onDragOver( event ) {
-		event.preventDefault();
```
Is it ok to remove that? What was  the reason it existed in the first place?

chriskmnds commented on components/drop-zone/provider.js:
```diff
@@ -133,11 +132,15 @@ class DropZoneProvider extends Component {
 		} );
+
+		event.preventDefault();
+		event.stopPropagation();
```
Reminder to check again and probably remove this last line. No need to stop propagation here.

nb replied:
Any decision here?

chriskmnds committed:
Merge branch 'master' into fix/dropzone-provider

nb commented on components/drop-zone/provider.js:
```diff
@@ -162,14 +165,18 @@ class DropZoneProvider extends Component {
 		// This seemingly useless line has been shown to resolve a Safari issue
 		// where files dragged directly from the dock are not recognized
 		event.dataTransfer && event.dataTransfer.files.length; // eslint-disable-line no-unused-expressions
+
```
Do we need those empty lines?

chriskmnds replied:
Yes :p

nb commented:
Good catch and works well for me. Added @youknowriad as a review to confirm whether we need `isDraggingOverDocument` and for another sanity check.

chriskmnds committed:
Merge branch 'master' into fix/dropzone-provider

chriskmnds committed:
Fix for Issue 4329: Removed inclusion of stopPropagation call in the dragover callback.

chriskmnds commented:
Awesome @nb! Thanks, I also pinged @youknowriad in the comments to shed some light on that missing `onVerifyValidTransfer` property.

youknowriad commented on components/drop-zone/provider.js:
```diff
@@ -76,10 +71,6 @@ class DropZoneProvider extends Component {
-		event.preventDefault();
-	}
-
 	toggleDraggingOverDocument( event ) {
```
This function is called so many times when dragging right now, do you think we should throttle it?

chriskmnds replied:
Indeed. Great idea Riad. I am looking into this, currently trying something out with lodash. Sorry for the delayed response, I was just away since yesterday.

chriskmnds replied:
Hey Riad, I've finally added throttling at 300ms using lodash. It feels ok to me at this value, but please have a go and let me know what you think. Also, this is the best I could come up with code-wise to get the `event.preventDefault` call working. Tested it in Chrome/Firefox/Safari/IE11. :-)

youknowriad reviewed (commented):
So this change can create performance issues when dragging files in theory but I'm not seeing the performance issues we had when we first tried this.

I think we optimized the blocks to avoid rerendering without changes which makes this strategy possible now.

cc @jasmussen for sanity check (performance wise) but in general, this is looking good.

jasmussen commented:
In my quick tests here, dragging and dropping both single and multiple images into a long document caused no performance issues for me and was simply a delight.

chriskmnds commented:
> I think we optimized the blocks to avoid rerendering without changes which makes this strategy possible now.

Indeed. I noticed some smart checks in there to only update the dropzones that need to be updated. Last I remember it was I believe 2 state updates as we drag something over a block top to bottom. I will run a few more checks to be sure and clean up that `onVerifyValidTransfer` property.

chriskmnds committed:
Fix for Issue 4329: Cleanup - removed redundant check with undefined property onVerifyValidTransfer.

chriskmnds committed:
Fix for Issue 4329: Improved border overlap between dropzones.

- The top/bottom negative offsets of the block dropzones allow for the
  top/bottom borders shown when on hover to overlap, hence visualy appearing
  as one (or not changing). This was off by 1px, which this commit fixes.

chriskmnds commented:
@jasmussen when/if you get a chance, can you please take a look at my very last commit here? There is a brief description there - not sure if you noticed it, but if you look closer there is a 1px difference in overlap of the top/bottom blue borders of stacked dropzones. These are supposed to overlap exactly when you drag something from one block to the next. This seemingly simple change fixes it, but I can't determine what the cause was. Any ideas? shall we just call it fixed and not worry?

jasmussen commented:
Sure thing, I'll take a look first thing tomorrow morning.

chriskmnds commented:
Thanks :-)

jasmussen commented:
Checked out your branch, tried to drag and drop some images in, tried to create some reusable blocks, tried dragging and dropping again. I can't seem to find any _visual_ regressions in this change. If you see the same, then let's call it fixed for now, and we can revisit if any issues pop up.

youknowriad commented:
This comment https://github.com/WordPress/gutenberg/pull/4604#discussion_r164363566 doesn't seem resolved yet. Any thoughts about this?

chriskmnds commented:
Thanks @jasmussen ! I was just puzzled why I had to reduce the bottom offset by 1px to get the blue borders aligned when you drag across blocks -but yeah, it's such a minor issue, don't worry about it if it's looking fine. Thanks for checking 😄

chriskmnds committed:
Fix for Issue 4329: Added throttling of 500 ms to dropzone provider's dragover listener.

chriskmnds committed:
Fix for Issue 4329: Reduced throttling to 300 ms.

chriskmnds committed:
Fix for Issue 4329: Cleanup - linting errors.

chriskmnds committed:
Merge branch 'master' into fix/dropzone-provider

youknowriad commented on components/drop-zone/provider.js:
```diff
@@ -17,6 +17,13 @@ class DropZoneProvider extends Component {
 
+		const throttledAction = throttle( this.toggleDraggingOverDocument, 500 );
+
+		this.dragOverListener = ( event ) => {
```
Minor: I guess it's just a preference of mine but what do you think about something like this:

```js
this.toggleDraggingOverDocument = throttle( this.toggleDraggingOverDocument.bind( this ), 500 );
this.dragOverListener = this.dragOverListener.bind( this );
```

and move the `dragOverListener` declaration as a regular method (not defined in the constructor).

youknowriad replied:
Also wondering if `500` is not too high :) maybe 200?

chriskmnds replied:
Indeed, `200` is a lot better.

> Minor: I guess it's just a preference of mine but what do you think about something like this:...

Definitely. That is a lot cleaner/better. Just updated now. :-) 

Also, you'll notice I make the call to the throttled method as `this.toggleDraggingOverDocument.call( this, event );` - it's not necessary - we can just `this.toggleDraggingOverDocument( event )` - lodash uses the last `this` value on each call. I don't know why but I feel safer to have it this way instead? Should I just revert to the normal invocation?

chriskmnds replied:
Sorry this is going back/forth so many times... :-)

youknowriad replied:
No worries, thanks for handling all these details :)

I'd be in favor of a regular `this.toggleDraggingOverDocument( event )` 

I'll give it a last spin tomorrow morning and approve

chriskmnds replied:
Awesome! Updated. Let me know if you spot anything else :-)

chriskmnds committed:
Fix for Issue 4329: Cleanup/refactored after Riad's comments + lowered throttle interval to 200.

chriskmnds committed:
Fix for Issue 4329: Cleanup/refactored after review.

youknowriad reviewed (approved):
Nice work here 👍 Thanks for addressing all the concerns

youknowriad commented:
Hey @chriskmnds I'm seeing a small bug on master where there's a lingering placeholder at the end of the post when drag and dropping a file.

I wonder if it's related to this PR

chriskmnds commented:
Hey @youknowriad I notice that too. It doesn't look like it's related to this PR. I've checked the commit before it (d8e13bf6dd65bba08ecbcc5d923453022fb2f7b4) and I can still see the lingering placeholder. This is what you are referring to right:

![placeholder-bottom-1](https://user-images.githubusercontent.com/1705499/35692086-0cb55056-0783-11e8-9577-fb16bcd3a3ca.png)

or

![placeholder-bottom-2](https://user-images.githubusercontent.com/1705499/35692102-1627acc4-0783-11e8-85be-2203566da914.png)

I sense this is related to the latest changes with the sibling inserter. I can look into this first thing tomorrow in any case. Shall I create a ticket for it?  :-)
